sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkCharVarcharUtils.scala

```diff
@@ -54,7 +54,8 @@ trait SparkCharVarcharUtils {
       StructType(fields.map { field =>
         field.copy(dataType = replaceCharVarcharWithString(field.dataType))
       })
-    case CharType(_) | VarcharType(_) if !SqlApiConf.get.preserveCharVarcharTypeInfo => StringType
+    case st: StringType if !SqlApiConf.get.preserveCharVarcharTypeInfo =>
+      new StringType(st.collationId, st.constraint)
```

Comment:
Please, could you explain why this change is necessary? It does not seem anything intuitive why we do this, and also, it is a risky change, as we are chaning Char/Varchar from StringType object to something that previously returned StringType object and did not have any relation to the constraint. We should in general if we can, avoid changes like this.